### DISCUSSION TITLE
Go Templating

### DISCUSSION THREAD

**Discussion by wasabii** [dotnet/dotnet-operator-sdk#404] [Ideas & General Discussion]
So, this is more of a discussion topic. I am aware there is no easy answer for this.

There is a general pattern of developing Kubernetes operators that generate other objects using Golang templates embedded as strings within a CRD. An example is the 'ExternalSecret' type, which supports pulling values from Azure Keyvault, etc, into K8S Secrets.

The 'ExternalSecret' CRD, for example, allows you to simply put a Golang template into the body of a Spec property, and the operator executes that template against the external secret data, allowing you to format the contents of the external secret.

https://external-secrets.io/v0.4.4/guides-templating/

Since K8s is so go-y, this makes good sense. It fits nicely into the eco system.

However, this presents an issue when using The DotNet operator! We don't have access to Go-lang templates! What to do!

**Comment by wasabii** [comment]
I happen to have a need for this currently in my operator. So my approach right now is going to be to include a Handlebars template, since there is a well-designed and maintained Handlebars.NET package. But this is unfortunate, as my CRDs won't fit into the overall K8s ecosystem as cleanly.

The proper solution would probably be a DotNet native implementation of Go templates.

A less good solution would be invoking the Go lang template engine from within a DotNet operator.

**Reply by buehler** [reply]
So you basically "just need" a templating engine and put the template into the CRD to create another resource right? Then we do not need go templates, there are - as you mentioned - plenty of templating engines (even small ones) that can do the job. Since you are the one that creates the CRD and the operator, you may select the appropriate templating engine. 

Or do I misunderstand the way that it works in go?

**Reply by wasabii** [reply]
All of that is true. I can use whatever template engine I want. The only concern I bring up is that everybody else in the K8s ecosystem uses the Go templating engine, for obvious reasons. And so, an operator written in .NET would benefit from being able to use the same templating language in order to fit in.

**Reply by wasabii** [reply]
This is no doubt beyond the scope of this project. But it does go towards how this project fits into the wider K8s ecosystem. It would benefit authors of operators using the DotNet SDK to be able to expose templating to their customers in a format they are used to. So, this project would benefit from somebody else writing a GO template engine for .NET.
